src/node.cc

```diff
@@ -1303,86 +1303,187 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
+  delete private_;
+}
+
+InternalCallbackScope::InternalCallbackScope(Environment* env,
+                                             Local<Object> object,
+                                             const async_context& asyncContext)
+  : env_(env),
+    async_context_(asyncContext),
+    callback_scope_(env) {
+  CHECK(!object.IsEmpty());
+  if (object->IsObject())
+    object_ = object.As<Object>();
+
+  // If you hit this assertion, you forgot to enter the v8::Context first.
+  CHECK_EQ(env->context(), env->isolate()->GetCurrentContext());
+
   if (env->using_domains()) {
-    CHECK(recv->IsObject());
-    disposed_domain = DomainEnter(env, object);
-    if (disposed_domain) return Undefined(env->isolate());
+    CHECK(object->IsObject());
+    failed_ = DomainEnter(env, object_);
+    if (failed_) {
+      failure_return_value_ = Undefined(env->isolate());
+      return;
+    }
   }
 
-  MaybeLocal<Value> ret;
+  if (asyncContext.async_id != 0) {
+    // No need to check a return value because the application will exit if
+    // an exception occurs.
+    AsyncWrap::EmitBefore(env, asyncContext.async_id);
+  }
 
-  {
-    AsyncHooks::ExecScope exec_scope(env, asyncContext.async_id,
-                                     asyncContext.trigger_async_id);
+  env->async_hooks()->push_ids(async_context_.async_id,
+                               async_context_.trigger_async_id);
+  pushed_ids_ = true;
+}
 
-    if (asyncContext.async_id != 0) {
-      // No need to check a return value because the application will exit if
-      // an exception occurs.
-      AsyncWrap::EmitBefore(env, asyncContext.async_id);
-    }
+InternalCallbackScope::~InternalCallbackScope() {
+  Close();
+}
 
-    ret = callback->Call(env->context(), recv, argc, argv);
+void InternalCallbackScope::Close() {
+  if (closed_) return;
+  closed_ = true;
 
-    if (ret.IsEmpty()) {
-      // NOTE: For backwards compatibility with public API we return Undefined()
-      // if the top level call threw.
-      return callback_scope.in_makecallback() ?
-          ret : Undefined(env->isolate());
-    }
+  if (try_catch_ != nullptr && try_catch_->real_try_catch.HasCaught())
+    failed_ = true;
 
-    if (asyncContext.async_id != 0) {
-      AsyncWrap::EmitAfter(env, asyncContext.async_id);
-    }
+  delete try_catch_;
+  try_catch_ = nullptr;
+
+  if (pushed_ids_)
+    env_->async_hooks()->pop_ids(async_context_.async_id);
+
+  if (failed_) return;
+
+  if (async_context_.async_id != 0) {
+    AsyncWrap::EmitAfter(env_, async_context_.async_id);
   }
 
-  if (env->using_domains()) {
-    disposed_domain = DomainExit(env, object);
-    if (disposed_domain) return Undefined(env->isolate());
+  if (env_->using_domains()) {
+    failed_ = DomainExit(env_, object_);
+    if (failed_) {
+      failure_return_value_ = Undefined(env_->isolate());
+      return;
+    }
   }
 
-  if (callback_scope.in_makecallback()) {
-    return ret;
+  if (IsInnerMakeCallback()) {
+    return;
```

Comment:
this is purely my personal opinion of having balance between what runs entering and exiting the scope, but I feel that from here down should be put into it's own method (e.g. `PostProcessing()`) then at the end of `InternalMakeCallback()` do the `IsInnerMakeCallback()` check after `scope.Close()` then run it.

again, this isn't a technical issue, but it jumps at me every time I see it.

Reply:
I’m not 100 % following – you mean, move this from `scope.Close()` to `InternalMakeCallback()`? That would defeat the whole purpose of exposing `CallbackScope`, though…